### DISCUSSION TITLE
API for setting unencrypted cookies client-side

### DISCUSSION THREAD

**Discussion by nathggns** [remix-run/remix#9637] [Proposals]
Currently, the only way to set a cookie in a Remix app is using the `Set-Cookie` header, because the cookie library cannot run on the client. However, Remix has made specific decisions about the encoding of its cookies, which has to be manually replicated if you want to set the cookie on client-side (using `document.cookie`).

An example of this might be a cookie related to app layout – you can update this dynamically (i.e, resizing a sidebar) on the client, and on refresh you need to render server side using this information if you want to avoid content jump on client side render. This is typically done using a cookie (see https://github.com/bvaughn/react-resizable-panels?tab=readme-ov-file#how-can-i-use-persistent-layouts-with-ssr ), but to do this you have to manually copy how Remix encodes cookies. 

This is obviously not ideal and is brittle if remix were to ever change this encoding for whatever reason. 

The code encoding/decoding cookies is not exported and is within the `remix-server-runtime` https://github.com/remix-run/remix/blob/main/packages/remix-server-runtime/cookies.ts

**Comment by kiliman** [comment]
You can use standard cookies. Here's an example that uses `js-cookie` in the browser and `cookie` on the server.
```ts
import Cookies from "js-cookie";
import cookie from "cookie";

export function loader({ request }: LoaderFunctionArgs) {
  // read server cookies
  const cookies = cookie.parse(request.headers.get("cookie") ?? "");
  return json({ cookies });
}
export default function Index() {
  const data = useLoaderData<typeof loader>();
  const { revalidate } = useRevalidator();
  const setTheme = (theme: string) => {
    // set client cookie
    Cookies.set("theme", theme);
    console.log("setting theme", document.cookie);
    revalidate();
  };
  //...
}
```
https://github.com/kiliman/remix-js-cookie

**Reply by alexanderson1993** [reply]
It's worth pointing out that Remix's implementation is specifically designed to make sure cookies aren't being spoofed by the client. In other words, it's trying to make it impossible to update `document.cookie`. OP uses the word "encoding" in the post to describe what Remix does to cookies. That's not quite right - Remix is cryptographically signing the cookies with a key that only the server has access to to ensure that only the server can set cookies.

It's a security and consistency measure that should be used by most apps. But if you really do want to modify cookies client-side, the solution Michael posted above is what you want to do.

**Reply by kiliman** [reply]
Yes, I would only set client-side cookies for non-critical data like themes or persisting layout info, etc.

**Reply by nathggns** [reply]
Remix does not cryptographically sign all cookies. `createCookie` by default does not sign cookies –you must pass an array of secrets to turn on signing. This is not what I'm referring to.

Not all cookies are "secrets" and so some you will not need to sign. Some of these you will want to be able to update from the client without a network request (layout is a good example of this). 

You can use of course use other packages for setting/parsing cookies. But if you want to use `createCookie` on the server side (the library built into Remix), there is no "official" way to support updating this for unsigned cookies without a network request.

I don't want to add new dependencies (although its an implementation detail that Remix currently uses the `cookie` package) for this or to replicate the way Remix encodes cookie values. I think Remix should support updating unsigned cookies from the client as a first class feature. 

Even with unsigned cookies, Remix is still encoding the cookie value using this combo – `btoa(myUnescape(encodeURIComponent(JSON.stringify(value))))` – which is then further serialized by the `cookie` package.

This is what I don't want to replicate within my own code. 

See https://github.com/remix-run/remix/blob/main/packages/remix-server-runtime/cookies.ts if you want to see what's going on.

It absolutely is correct to say remix is encoding cookies. It can also sign them for you, but that's a different feature and not one I'm referring to in this discussion.

**Reply by sergiodxa** [reply]
I think the cookies file in the `@remix-run/server-runtime` package can actually runs client-side, but you would need to do what `@remix-run/node` and `@remix-run/cloudflare` and provide a sign and unsign functions https://github.com/remix-run/remix/blob/main/packages/remix-cloudflare/crypto.ts, the Cloudflare version in particular should work client-side since it uses atob, btoa, Web Crypto and TextEncoder which are all web apis.

Once you have your sign and unsign functions, you can do https://github.com/remix-run/remix/blob/e324b3e08c83f561f386bdc4a836308cd5249271/packages/remix-cloudflare/implementations.ts#L10 to create a new createCookie function from the createCookieFactory.

**Reply by sergiodxa** [reply]
You can try to import `createCookie` from `@remix-run/cloudflare` and see if it just works.

**Comment by NGimbal** [comment]
So I was also trying to solve this problem and this discussion was very helpful. I inlined some of the the @remix-run/cloudflare code to create a `createClientCookie.ts` utility, it seems to be working. Let me know if you see any issues / ways this could be improved.

```typescript
// creatClientCookie.ts
import {
	createCookieFactory,
	type SignFunction,
	type UnsignFunction,
} from '@remix-run/server-runtime'

const encoder = new TextEncoder()

export const sign: SignFunction = async (value, secret) => {
	let key = await createKey(secret, ['sign'])
	let data = encoder.encode(value)
	let signature = await crypto.subtle.sign('HMAC', key, data)
	let hash = btoa(String.fromCharCode(...new Uint8Array(signature))).replace(
		/=+$/,
		'',
	)

	return value + '.' + hash
}

export const unsign: UnsignFunction = async (signed, secret) => {
	let index = signed.lastIndexOf('.')
	let value = signed.slice(0, index)
	let hash = signed.slice(index + 1)

	let key = await createKey(secret, ['verify'])
	let data = encoder.encode(value)
	let signature = byteStringToUint8Array(atob(hash))
	let valid = await crypto.subtle.verify('HMAC', key, signature, data)

	return valid ? value : false
}

async function createKey(
	secret: string,
	usages: CryptoKey['usages'],
): Promise<CryptoKey> {
	let key = await crypto.subtle.importKey(
		'raw',
		encoder.encode(secret),
		{ name: 'HMAC', hash: 'SHA-256' },
		false,
		usages,
	)

	return key
}

function byteStringToUint8Array(byteString: string): Uint8Array {
	let array = new Uint8Array(byteString.length)

	for (let i = 0; i < byteString.length; i++) {
		array[i] = byteString.charCodeAt(i)
	}

	return array
}

export const createClientCookie = createCookieFactory({ sign, unsign })
```

Usage with react-resizable-panel:
```typescript
async function handleLayout(sizes: number[]) {
  const cookie = await layoutCookie.serialize(sizes)
  document.cookie = cookie
}

...
return (
  <PanelGroup
    onLayout={handleLayout}
  >
...
```
